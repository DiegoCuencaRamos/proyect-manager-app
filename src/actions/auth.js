import { auth, GoogleAuthProvider, FacebookAuthProvider, TwitterAuthProvider, GithubAuthProvider } from "../firebase"

// LOGIN
const login = (uid) => ({
    type: 'auth/login',
    payload: uid
})

// Login-middleware
const startGoogleLogin = () => {
    return (dispatch, getState) => {
        return auth.signInWithRedirect(GoogleAuthProvider)
            .catch(e => console.log(e.message))
    }
}

const startFacebookLogin = () => {
    return (dispatch, getState) => {
        return auth.signInWithRedirect(FacebookAuthProvider)
            .catch(e => console.log(e.message))
    }
}

const startTwitterLogin = () => {
    return (dispatch, getState) => {
        return auth.signInWithRedirect(TwitterAuthProvider)
            .catch(e => console.log(e.message))
    }
}

const startGithubLogin = () => {
    return (dispatch, getState) => {
        return auth.signInWithRedirect(GithubAuthProvider)
            .catch(e => console.log(e.message))
    }
}

const startEmailLogin = (email, password, setErrorMessage) => {
    return (dispath, getState) => {
        return auth.signInWithEmailAndPassword(email, password)
            .catch(e => {
                console.log(e.message)
                if(e.message === 'The password is invalid or the user does not have a password.') {
                    setErrorMessage('he password is invalid')
                } else {
                    setErrorMessage(e.message)
                }  
            })
    }
}

const startEmailSignUp = (email, password, setErrorMessage) => {
    return (dispatch, getState) => {
        return auth.createUserWithEmailAndPassword(email, password)
            .catch(e => {
                console.log(e.message)
                setErrorMessage(e.message)
            })
    }
}

// LOGOUT
const logout = () => ({
    type: 'auth/logout'
})

const startLogout = () => {
    return (dipatch, getState) => {
        return auth.signOut()
    }
}

export {
    login,
    startGoogleLogin,
    startFacebookLogin,
    startTwitterLogin,
    startGithubLogin,
    startEmailLogin,
    startEmailSignUp,
    logout,
    startLogout
}