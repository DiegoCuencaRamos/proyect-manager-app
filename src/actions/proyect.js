import database from '../firebase'
import moment from 'moment'

// 1. Add proyect
const addProyect = (proyect) => ({
    type: 'proyect/addProyect',
    payload: proyect
})

const startAddProyect = (proyectData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        const {
            name = '', 
            description = '', 
            status = '', 
            invoice = undefined,
            startDate = moment().startOf('month'),
            endDate = moment().endOf('month')

        } = proyectData
        const proyect = { name, description, status, invoice, startDate, endDate }

        return database.ref(`/users/${uid}/proyects`).push(proyect)
            .then(ref => {
                dispatch(addProyect({
                    id: ref.key,
                    ...proyect
                }))
            })
    }
}

// 2. Edit proyect
const editProyect = (id, updates) => ({
    type: 'proyect/editProyect',
    payload: {
        id,
        updates
    }
})

const startEditProyect = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid

        return database.ref(`/users/${uid}/proyects/${id}`).update(updates)
            .then(() => {
                dispatch(editProyect(id, updates))
            })
    }
}

// 3. Remove proyect
const removeProyect = (id) => ({
    type: 'proyect/removeProyect',
    payload: id
})

const startRemoveProyect = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid

        return database.ref(`/users/${uid}/proyects/${id}`).remove()
            .then(() => {
                dispatch(removeProyect(id))
            })
    }
}

// 4. Set proyect
const setProyects = (proyects) => ({
    type: 'proyect/setProyects',
    payload: proyects
})

const startSetProyects = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid

        return database.ref(`/users/${uid}/proyects`).once('value')
            .then(dataSnapshot => {
                const proyects = []

                dataSnapshot.forEach(childDataSnapshot => {
                    proyects.push({
                        id: childDataSnapshot.key,
                        ...childDataSnapshot.val() 
                    })
                })

                dispatch(setProyects(proyects))
            })
    }
}

export { 
    addProyect,
    startAddProyect, 
    editProyect, 
    startEditProyect, 
    removeProyect, 
    startRemoveProyect, 
    setProyects, 
    startSetProyects 
}