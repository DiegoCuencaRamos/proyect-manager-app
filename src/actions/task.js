import database from '../firebase'

// 1. Add task
const addTask = (task) => ({
    type: 'task/addTask',
    payload: task
})

const startAddTask = (taskData = {}) => {
    return (dispatch, getState) => {
        const {
            proyectId = '', 
            name = '', 
            description = '', 
            status = ''
        } = taskData
        const task = { proyectId, name, description, status }

        return database.ref('/tasks').push(task)
            .then(ref => {
                dispatch(addTask({
                    id: ref.key,
                    ...task
                }))
            })     
    }
}

// 2. Edit task
const editTask = (id, updates) => ({
    type: 'task/editTask',
    payload: {
        id,
        updates
    }
})

const startEditTask = (id, updates) => {
    return (dispatch, getState) => {

        return database.ref(`/tasks/${id}`).update(updates)
            .then(() => {
                dispatch(editTask(id, updates))
            })
    }
}

// 3. Remove task
const removeTask = (id) => ({
    type: 'task/removeTask',
    payload: id
})

const startRemoveTask = (id) => {
    return (dispatch, getState) => {

        return database.ref(`/tasks/${id}`).remove()
            .then(() => {
                dispatch(removeTask(id))
            })
    }
}

// 4. Set tasks
const setTasks = (tasks) => ({
    type: 'task/setTasks',
    payload: tasks
})

const startSetTasks = () => {
    return (dispatch, getState) => {

        return database.ref(`/tasks`).once('value')
            .then(dataSnapshot => {
                const tasks = []

                dataSnapshot.forEach(dataChildSnapshot => {
                    tasks.push({
                        id: dataChildSnapshot.key,
                        ...dataChildSnapshot.val()
                    })
                })

                dispatch(setTasks(tasks))
            })
    }
}

export { 
    addTask, 
    startAddTask,
    editTask, 
    startEditTask,
    removeTask,
    startRemoveTask,
    setTasks,
    startSetTasks
}