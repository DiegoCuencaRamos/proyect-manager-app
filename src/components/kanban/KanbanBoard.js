import React from 'react'
import KanbanBoardHeader from './KanbanBoardHeader';
import KanbanBoardBody from './KanbanBoardBody';

const KanbanBoard = () => {
    // Variables
    const statuses =  ['todo', 'doing', 'done']
     
    // Render
    return (
        <section className="kanban">
            <div className="container">
                {statuses.map((status, index) => (
                    <div key={index} className={`kanban__board--${status}`}>
                        <KanbanBoardHeader status={status} />
                        <KanbanBoardBody status={status} />
                    </div> 
                ))}
            </div>
        </section>   
    )
}

export default KanbanBoard