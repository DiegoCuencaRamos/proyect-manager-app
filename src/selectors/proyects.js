const getFilteredProyects = (proyects, textFilter, sortBy) => {
    return proyects
        // Get filtered proyects by textFilter
        .filter(proyect => proyect.name.toLowerCase().includes(textFilter.toLowerCase()))
        // Order proyects by criteria
        .sort((a, b) => {
            if(sortBy === 'name') {
                return a.name.toLowerCase() > b.name.toLowerCase() ? 1 : -1
            }
        })
    
}

export default getFilteredProyects