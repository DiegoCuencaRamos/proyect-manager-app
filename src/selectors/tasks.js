const getVisibleTasks = (proyectId, tasks, { searchText, sortBy }) => {
    return tasks
        // Get filtered tasks by proyect
        .filter(task => task.proyectId === proyectId)
        // Get filtered tasks by searchText
        .filter(task => task.name.toLowerCase().includes(searchText.toLowerCase()))
        // Order tasks by criteria
        .sort((a, b) => {
            if(sortBy === 'name') {
                return a.name.toLowerCase() > b.name.toLowerCase() ? 1 : -1
            }
        }) 
}

export default getVisibleTasks