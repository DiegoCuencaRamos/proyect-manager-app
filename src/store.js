import { createStore } from 'redux'
import rootReducer from './reducer'

const enhancer = window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()

const store = createStore(
    rootReducer,
    /* preloadedState. */
    enhancer,
)

export default store