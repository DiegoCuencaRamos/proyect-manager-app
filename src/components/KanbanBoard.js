import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { setProyectId } from '../actions/id'

const KanbanBoard = ({ isProyect }) => {
    // Variables
    const statuses =  ['todo', 'doing', 'done']
    const items = useSelector(state => state[ isProyect ? 'proyects' : 'tasks' ])
    // Events
    let onItemClicked
    if (isProyect) {
        const dispatch = useDispatch()
        onItemClicked = (e) => {
            const id = e.target.dataset.id
            dispatch(setProyectId(id))
        }
    }
    // Rendering
    return (
        <section className="kanban">
            <div className="container">
                {
                    statuses.map((status, index) => (
                        <div 
                            key={index}
                            className={`kanban__board--${status}`}
                        >
                            <div className="kanban__thead">
                                <h3>{status}</h3>
                            </div>
                            <div className="kanban__tbody">
                                <Link 
                                    key={index}
                                    to={isProyect ? `/add-proyect` : `/add-task`}
                                    className="kanban__item"
                                >
                                    <p className="kanban__add-plus">+</p>
                                    <p className="kanban__add-text">{`Create new ${isProyect ? 'proyect' : 'task'} here`}</p>
                                </Link>

                                    {
                                        items
                                            .filter(item => item.status === status)
                                            .map(({ id, name, description, invoice = undefined }) => (
                                                <Link 
                                                    key={id}
                                                    to={isProyect ? `/proyect/${id}` : `/task/${id}`}
                                                    className="kanban__item"
                                                    data-id={id}
                                                    onClick={onItemClicked}
                                                >
                                                    <p className="kanban__item__title">{name}</p>
                                                    <p className="kanban__item__description">{description}</p>
                                                    <p className="kanban__item__invoice">{invoice}</p>
                                                </Link>
                                            ))
                                    }
                            </div>
                        </div>
                    ))
                }
            </div>
        </section>
        
    )
}

export default KanbanBoard