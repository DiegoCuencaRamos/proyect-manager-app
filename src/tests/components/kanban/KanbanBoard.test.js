import React from 'react'
import { mount } from 'enzyme'
import { MemoryRouter } from 'react-router'
import { Provider } from 'react-redux'
import store from '../../../store'
import DashboardContext from '../../../contexts/dashboard-context'
import KanbanBoard from '../../../components/kanban/KanbanBoard'
import { proyects } from '../../fixures/proyects'
import { addProyect } from '../../../actions/proyect'
import { tasks } from '../../fixures/tasks'
import { addTask } from '../../../actions/task'

test('Should render KanbanBoard correctly', () => {
    const wrapper = mount(
        <MemoryRouter>
            <Provider store={store}>
                <KanbanBoard />
            </Provider>
        </MemoryRouter>
    )
    expect(wrapper).toMatchSnapshot()
})

test('Should render KanbanBoard with proyects', () => {
    const isProyect = true

    proyects.forEach(proyect => {
        store.dispatch(addProyect(proyect))
    })

    const wrapper = mount(
        <MemoryRouter>
            <Provider store={store}>
                <DashboardContext.Provider value={{ isProyect }}>
                    <KanbanBoard />
                </DashboardContext.Provider>
            </Provider>
        </MemoryRouter>
    )
})

