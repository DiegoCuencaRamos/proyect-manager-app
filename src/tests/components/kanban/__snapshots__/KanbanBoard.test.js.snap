// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render KanbanBoard correctly 1`] = `
<MemoryRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "zhw7qw",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "zhw7qw",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <KanbanBoard>
        <section
          className="kanban"
        >
          <div
            className="container"
          >
            <div
              className="kanban__board--todo"
              key="0"
            >
              <KanbanBoardHeader
                status="todo"
              >
                <div
                  className="kanban__thead"
                >
                  <h3>
                    todo
                  </h3>
                </div>
              </KanbanBoardHeader>
              <KanbanBoardBody
                status="todo"
              >
                <div
                  className="kanban__tbody"
                >
                  <Link
                    className="kanban__item"
                    to="/add-task"
                  >
                    <LinkAnchor
                      className="kanban__item"
                      href="/add-task"
                      navigate={[Function]}
                    >
                      <a
                        className="kanban__item"
                        href="/add-task"
                        onClick={[Function]}
                      >
                        <p
                          className="kanban__add-plus"
                        >
                          +
                        </p>
                        <p
                          className="kanban__add-text"
                        >
                          Create new task here
                        </p>
                      </a>
                    </LinkAnchor>
                  </Link>
                </div>
              </KanbanBoardBody>
            </div>
            <div
              className="kanban__board--doing"
              key="1"
            >
              <KanbanBoardHeader
                status="doing"
              >
                <div
                  className="kanban__thead"
                >
                  <h3>
                    doing
                  </h3>
                </div>
              </KanbanBoardHeader>
              <KanbanBoardBody
                status="doing"
              >
                <div
                  className="kanban__tbody"
                >
                  <Link
                    className="kanban__item"
                    to="/add-task"
                  >
                    <LinkAnchor
                      className="kanban__item"
                      href="/add-task"
                      navigate={[Function]}
                    >
                      <a
                        className="kanban__item"
                        href="/add-task"
                        onClick={[Function]}
                      >
                        <p
                          className="kanban__add-plus"
                        >
                          +
                        </p>
                        <p
                          className="kanban__add-text"
                        >
                          Create new task here
                        </p>
                      </a>
                    </LinkAnchor>
                  </Link>
                </div>
              </KanbanBoardBody>
            </div>
            <div
              className="kanban__board--done"
              key="2"
            >
              <KanbanBoardHeader
                status="done"
              >
                <div
                  className="kanban__thead"
                >
                  <h3>
                    done
                  </h3>
                </div>
              </KanbanBoardHeader>
              <KanbanBoardBody
                status="done"
              >
                <div
                  className="kanban__tbody"
                >
                  <Link
                    className="kanban__item"
                    to="/add-task"
                  >
                    <LinkAnchor
                      className="kanban__item"
                      href="/add-task"
                      navigate={[Function]}
                    >
                      <a
                        className="kanban__item"
                        href="/add-task"
                        onClick={[Function]}
                      >
                        <p
                          className="kanban__add-plus"
                        >
                          +
                        </p>
                        <p
                          className="kanban__add-text"
                        >
                          Create new task here
                        </p>
                      </a>
                    </LinkAnchor>
                  </Link>
                </div>
              </KanbanBoardBody>
            </div>
          </div>
        </section>
      </KanbanBoard>
    </Provider>
  </Router>
</MemoryRouter>
`;
